antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/InterpreterDataReader.kt

```diff
@@ -1,129 +1,163 @@
 // Copyright 2017-present Strumenta and contributors, licensed under Apache 2.0.
 // Copyright 2024-present Strumenta and contributors, licensed under BSD 3-Clause.
-
 package org.antlr.v4.kotlinruntime.misc
 
+import com.strumenta.antlrkotlin.runtime.LineReader
 import org.antlr.v4.kotlinruntime.Vocabulary
+import org.antlr.v4.kotlinruntime.VocabularyImpl
 import org.antlr.v4.kotlinruntime.atn.ATN
+import org.antlr.v4.kotlinruntime.atn.ATNDeserializer
 
-// TODO(Edoardo): this one requires reading files
-// A class to read plain text interpreter data produced by ANTLR.
+/**
+ * A class to read plain text interpreter data produced by ANTLR.
+ */
 public object InterpreterDataReader {
-  public class InterpreterData {
-    internal var atn: ATN? = null
-    internal var vocabulary: Vocabulary? = null
-    internal var ruleNames: MutableList<String>? = null
-    internal var channels: MutableList<String>? = null // Only valid for lexer grammars.
-    internal var modes: MutableList<String>? = null // ditto
-  }
+  public data class InterpreterData(
+    public val atn: ATN,
+    public val vocabulary: Vocabulary,
+    public val ruleNames: List<String>,
+    public val channels: List<String>?, // Only valid for lexer grammars
+    public val modes: List<String>?, // ditto
+  )
+
+  /**
+   * The structure of the data file is very simple.
+   * Everything is line based with empty lines separating the different parts.
+   *
+   * For lexers the layout is:
+   *
+   * ```
+   * token literal names:
+   * ...
+   *
+   * token symbolic names:
+   * ...
+   *
+   * rule names:
+   * ...
+   *
+   * channel names:
+   * ...
+   *
+   * mode names:
+   * ...
+   *
+   * atn:
+   * <a single line with comma separated int values enclosed in a pair of square brackets>
+   * ```
+   *
+   * Data for a parser does not contain channel and mode names.
+   */
+  public fun parse(reader: LineReader): InterpreterData {
+    var line = reader.readLineOrNull()
+
+    if (line != "token literal names:") {
```

Comment:
Would it make sense to create an helper function, as we replicate the same structure multiple times. Or is this done to be aligned with the Java runtime line by line?

Reply:
Yup I don't want to make it different than the Java one. Mostly to avoid confusion if we ever have to port changes.